File: backend/app/schemas/stock.py
"""API response shapes for the /stocks endpoints.

These are the public contract the frontend consumes. They're separate from the
provider DTOs on purpose: the wire format the app exposes shouldn't be coupled
to whatever a vendor happens to return.
"""

from datetime import datetime

from pydantic import BaseModel


class SearchResultOut(BaseModel):
    symbol: str
    name: str | None = None
    exchange: str | None = None
    type: str | None = None


class StockOut(BaseModel):
    ticker: str
    company_name: str | None = None
    exchange: str | None = None
    sector: str | None = None
    currency: str | None = None
    metadata: dict = {}
    updated_at: datetime | None = None


class QuoteOut(BaseModel):
    ticker: str
    price: float
    previous_close: float | None = None
    change: float | None = None
    change_percent: float | None = None
    open: float | None = None
    day_high: float | None = None
    day_low: float | None = None
    volume: int | None = None
    market_cap: float | None = None
    fifty_two_week_high: float | None = None
    fifty_two_week_low: float | None = None
    currency: str | None = None
    exchange: str | None = None
    as_of: datetime


class CandleOut(BaseModel):
    timestamp: datetime
    open: float
    high: float
    low: float
    close: float
    adj_close: float | None = None
    volume: int | None = None


class HistoryOut(BaseModel):
    ticker: str
    range: str
    interval: str
    candles: list[CandleOut]

File: frontend/types/stock.ts
// Mirrors the backend Pydantic schemas in backend/app/schemas/stock.py.
// Keeping these in sync is manual for now; a future phase could codegen them
// from the OpenAPI spec FastAPI already exposes at /openapi.json.

export const RANGES = ["1d", "5d", "1m", "6m", "ytd", "1y", "5y"] as const;
export type Range = (typeof RANGES)[number];

export interface SearchResult {
  symbol: string;
  name: string | null;
  exchange: string | null;
  type: string | null;
}

export interface Stock {
  ticker: string;
  company_name: string | null;
  exchange: string | null;
  sector: string | null;
  currency: string | null;
  metadata: Record<string, unknown>;
  updated_at: string | null;
}

export interface Quote {
  ticker: string;
  price: number;
  previous_close: number | null;
  change: number | null;
  change_percent: number | null;
  open: number | null;
  day_high: number | null;
  day_low: number | null;
  volume: number | null;
  market_cap: number | null;
  fifty_two_week_high: number | null;
  fifty_two_week_low: number | null;
  currency: string | null;
  exchange: string | null;
  as_of: string;
}

export interface Candle {
  timestamp: string;
  open: number;
  high: number;
  low: number;
  close: number;
  adj_close: number | null;
  volume: number | null;
}

export interface History {
  ticker: string;
  range: Range;
  interval: string;
  candles: Candle[];
}
